src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1610,23 +1620,45 @@ bool tryScanRecordStart([NotNullWhen(true)] out SyntaxToken? keyword, out Syntax
                 return false;
             }
 
-            static TypeDeclarationSyntax constructTypeDeclaration(ContextAwareSyntax syntaxFactory, SyntaxList<AttributeListSyntax> attributes, SyntaxListBuilder modifiers, SyntaxToken keyword, SyntaxToken? recordModifier,
-                SyntaxToken name, TypeParameterListSyntax typeParameters, ParameterListSyntax? paramList, BaseListSyntax baseList, SyntaxListBuilder<TypeParameterConstraintClauseSyntax> constraints,
+            bool tryScanExtensionStart([NotNullWhen(true)] out SyntaxToken? implicitOrExplicitKeyword, [NotNullWhen(true)] out SyntaxToken? extensionKeyword)
+            {
+                // PROTOTYPE consider improving error recovery for `extension explicit` and `extension implicit`
+                if (this.CurrentToken.Kind is SyntaxKind.ExplicitKeyword or SyntaxKind.ImplicitKeyword &&
+                    this.PeekToken(1).ContextualKind == SyntaxKind.ExtensionKeyword)
+                {
+                    implicitOrExplicitKeyword = ConvertToKeyword(EatToken());
+                    extensionKeyword = ConvertToKeyword(EatToken());
+                    return true;
+                }
+
+                implicitOrExplicitKeyword = null;
+                extensionKeyword = null;
+                return false;
+            }
+
+            static TypeDeclarationSyntax constructTypeDeclaration(ContextAwareSyntax syntaxFactory, SyntaxList<AttributeListSyntax> attributes, SyntaxListBuilder modifiers,
+                SyntaxToken? firstKeyword, SyntaxToken? secondKeyword, SyntaxToken mainKeyword,
+                SyntaxToken name, TypeParameterListSyntax typeParameters, ParameterListSyntax? paramList, SyntaxToken? forKeyword, TypeSyntax? underlyingType,
+                BaseListSyntax baseList, SyntaxListBuilder<TypeParameterConstraintClauseSyntax> constraints,
                 SyntaxToken? openBrace, SyntaxListBuilder<MemberDeclarationSyntax> members, SyntaxToken? closeBrace, SyntaxToken semicolon)
             {
                 var modifiersList = (SyntaxList<SyntaxToken>)modifiers.ToList();
                 var membersList = (SyntaxList<MemberDeclarationSyntax>)members;
                 var constraintsList = (SyntaxList<TypeParameterConstraintClauseSyntax>)constraints;
-                switch (keyword.Kind)
+                switch (mainKeyword.Kind)
                 {
                     case SyntaxKind.ClassKeyword:
+                        RoslynDebug.Assert(firstKeyword is not null);
```

Comment:
it's confusing to me that we have mainKeyword and firstKeyword.  especially since we clearly see that mainKeyword here is the ClassKeyword.  but we then do *not* pass that in for `ClassDeclaration( ... keyword: ... )`